table/block_based/block_based_table_reader.cc

```diff
@@ -3053,4 +3050,34 @@ void BlockBasedTable::DumpKeyValue(const Slice& key, const Slice& value,
   out_stream << "  ------\n";
 }
 
+template <typename TBlocklike>
+Cache::CreateCallback BlockBasedTable::GetCreateCallback(
+    BlockType block_type) const {
+  // limit the number of captured variables of the lambda to 2, so for the most
+  // of the compilers, it won't dynamically allocate extra memory from heap
+  // (which could be relatively expensive for the read path).
+  // Note: don't just capture the reference to an obj on stack, as the callback
+  //  may out live the object on stack, as the user lib may copy the callback in
+  //  lookup().
+  // For long-term, if it does need to capture more variables, maybe only do
+  // that for secondary cache?
+  return [this, block_type](const void* buf, size_t size, void** out_obj,
```

Comment:
I don't believe this works, because there's no guarantee the table reader will be around for promotion from secondary to primary cache. Why? max_open_files > 0 could close table reader, or a different DB object using the same cache could mean old DB & table reader is closed.

Reply:
I guess if we assume the callback is only kept until an async Lookup is waited on, and if BlockBasedTable always waits on async Lookups (??? not sure), then I guess it's OK? Seems dangerous though if someone skips over wait, such as hitting an I/O error on one level of an iterator.

Reply:
I'm not sure about how long do we expect the callback should be valid. Is it fine to assume that the callback should only be called during the same read (like the async read)?

Reply:
The expectation is the higher level read operation (Get/MultiGet/Iter) will wait for the lookup to complete, so we can assume that the table reader is valid when the callback is called. I guess it can be enforced in ```Release``` if necessary, since Lookup returns a handle (even if the result is not immediately available) and it has to be released.

Reply:
I had hoped to make a "pending" async result into a distinct type, like FutureHandle, rather than Handle, so that we can enforce more discipline on its use. But tracking it as a Handle is rather deeply embedded in block_based_table_reader. :(